test/test_factbase.rb

```diff
@@ -219,4 +220,162 @@ def test_txn_with_rollback
     assert(!modified)
     assert_equal(0, fb.query('(always)').each.to_a.size)
   end
+
+  def test_concurrent_inserts
+    fb = Factbase.new
+    Threads.new(100).assert do
+      fact = fb.insert
+      fact.foo = 42
+      fact.bar = 49
+      fact.value = fact.foo * fact.bar
+    end
+    assert_equal(100, fb.size)
+    assert_equal(100, fb.query('(eq foo 42)').each.to_a.size)
+    assert_equal(100, fb.query('(eq bar 49)').each.to_a.size)
+    assert_equal(100, fb.query("(eq value #{42 * 49})").each.to_a.size)
+  end
+
+  def test_different_values_when_concurrent_inserts
+    fb = Factbase.new
+    Threads.new(100).assert do |i|
+      fb.insert.foo = i
+    end
+    assert_equal(100, fb.size)
+    Threads.new(100) do |i|
+      f = fb.query("(eq foo #{i})").each.to_a
+      assert_equal(1, f.count)
+      assert_equal(i, f.first.foo)
+    end
+  end
+
+  def test_different_properties_when_concurrent_inserts
+    skip
```

Comment:
@Suban05 what about this one? I believe, we need a puzzle for it to.

Reply:
@yegor256 I forgot about it, this test also fails, thanks